oauthlib/oauth2/rfc6749/grant_types/openid_connect.py

```diff
@@ -0,0 +1,408 @@
+
+from json import loads
+import logging
+
+import datetime
+
+from .base import GrantTypeBase
+from .authorization_code import AuthorizationCodeGrant
+from .implicit import ImplicitGrant
+from ..errors import InvalidRequestError, LoginRequired, ConsentRequired
+from ..request_validator import RequestValidator
+
+log = logging.getLogger(__name__)
+
+class OIDCNoPrompt(Exception):
+    """Exception used to inform users that no explicit authorization is needed.
+
+    Normally users authorize requests after validation of the request is done.
+    Then post-authorization validation is again made and a response containing
+    an auth code or token is created. However, when OIDC clients request
+    no prompting of user authorization the final response is created directly.
+
+    Example (without the shortcut for no prompt)
+
+    scopes, req_info = endpoint.validate_authorization_request(url, ...)
+    authorization_view = create_fancy_auth_form(scopes, req_info)
+    return authorization_view
+
+    Example (with the no prompt shortcut)
+    try:
+        scopes, req_info = endpoint.validate_authorization_request(url, ...)
+        authorization_view = create_fancy_auth_form(scopes, req_info)
+        return authorization_view
+    except OIDCNoPrompt:
+        # Note: Location will be set for you
+        headers, body, status = endpoint.create_authorization_response(url, ...)
+        redirect_view = create_redirect(headers, body, status)
+        return redirect_view
+    """
+
+    def __init__(self):
+        msg = ("OIDC request for no user interaction received. Do not ask user "
+               "for authorization, it should been done using silent "
+               "authentication through create_authorization_response. "
+               "See OIDCNoPrompt.__doc__ for more details.")
+        super(OIDCNoPrompt, self).__init__(msg)
+
+
+class OpenIDConnectBase(GrantTypeBase):
+
+    def __init__(self, request_validator=None):
+        self.request_validator = request_validator or RequestValidator()
+
+    def add_id_token(self, token, token_handler, request):
+        # Treat it as normal OAuth 2 auth code request if openid is not present
+        if not 'openid' in request.scopes:
+            return token
+
+        # Only add an id token on auth/token step if asked for.
+        if request.response_type and not 'id_token' in request.response_type:
+            return token
+
+        if not 'state' in token:
+            token['state'] = request.state
+
+        if request.max_age:
+            d = datetime.datetime.utcnow()
+            token['auth_time'] = d.isoformat("T") + "Z"
+
+        # TODO: acr claims (probably better handled by server code using oauthlib in get_id_token)
+
+        token['id_token'] = self.request_validator.get_id_token(token, token_handler, request)
+
+        if request.response_type == "id_token":
+            # implicit flow, only requesting id_token so don't return the access token
```

Comment:
This feels a bit hacky. If I understand correctly, in this scenario, we've already generated an access token, but then we're just throwing it away and not returning it to the user? I guess that's fine, but it seems like we should be able to do better and avoid creating access tokens we don't need. Can we dispatch on response_type earlier in the flow somewhere?